### PR TITLE
Fix mixing log/span decoration probes

### PR DESCRIPTION
**Pull request by jpbempel**
# What Does This Do
On same location, if we have different probes like log or span decoration probe with different characteristics (capture, snapshot, evaluateAt,...), instrumentation needs the correct definition to apply
 the shared CapturedContextInstrumentor.
We are creating a synthetic definition for applying instrumentation and merging all characteristics required for instrumentation.

# Motivation

# Additional Notes

Jira ticket: [DEBUG-2498]

[DEBUG-2498]: https://datadoghq.atlassian.net/browse/DEBUG-2498?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/DebuggerTransformer.java [hunk 1]
```diff
@@ -581,17 +585,73 @@ private List<ToInstrumentInfo> filterAndSortDefinitions(List<ProbeDefinition> de
...
+        if (definition instanceof ExceptionProbe) {
+          where = Where.convertLineToMethod(where, classFileLines);
+        }
+        hasLogProbe = true;
+        LogProbe logProbe = (LogProbe) definition;
+        captureSnapshot = captureSnapshot | logProbe.isCaptureSnapshot();
+        capture = mergeCapture(capture, logProbe.getCapture());
+      }
+      if (definition.getEvaluateAt() == MethodLocation.ENTRY
+          || definition.getEvaluateAt() == MethodLocation.DEFAULT) {
```
[the remark this answers is not in the document]
Comment (shatzi) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/DebuggerTransformer.java:611]: I think there should be a priority here.. if there are two probes one with DEFAULT and one with ENTRY.. the order of us going over the probe would have determine which one wins.

Does this even matter?

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/DebuggerTransformer.java [hunk 2]
```diff
@@ -581,17 +585,73 @@ private List<ToInstrumentInfo> filterAndSortDefinitions(List<ProbeDefinition> de
...
+          .build();
+    }
+    return SpanDecorationProbe.builder()
+        .probeId(probeId)
+        .where(where)
+        .evaluateAt(evaluateAt)
+        .build();
+  }
+
+  private LogProbe.Sampling mergeSampling(
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/DebuggerTransformer.java:631]: why do we care about merge the sampling? 

I thought we rate-limit each probe independently.. do we use the sampling of the reference definition?

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/DebuggerTransformerTest.java [hunk 3]
```diff
@@ -356,25 +356,36 @@ public void ordering() {
...
+    assertEquals(spanDecorationProbe, invocationOrder.get(3));
   }
 
   <T extends ProbeDefinition> T createMock(
       Class<T> clazz, List<ProbeDefinition> invocationOrder, String id) {
     ProbeDefinition mock = mock(clazz);
     doAnswer(
             invocation -> {
-              invocationOrder.add(mock);
+              // invocationOrder.add(mock);
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/DebuggerTransformerTest.java:388]: remove?

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/SpanDecorationProbeInstrumentationTest.java [hunk 4]
```diff
@@ -312,38 +314,51 @@ public void nullActiveSpan() throws IOException, URISyntaxException {
...
+            .capture(1, 50, 50, 10)
             .build();
     LogProbe logProbe2 =
         LogProbe.builder()
-            .probeId(PROBE_ID2)
+            .probeId(PROBE_ID4)
             .where(CLASS_NAME, "process")
             .captureSnapshot(true)
+            .sampling(10.0)
+            .capture(5, 200, 200, 30)
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/SpanDecorationProbeInstrumentationTest.java:343]: do we want to test that we merge the sampling and captures limit correctly?

**Review by shatzi — APPROVED** [review]

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/DebuggerTransformer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/DebuggerTransformer.java:611]: `DEFAULT` == `ENTRY` from the beginning.
The fact that by default, UI put `EXIT` is independent of what we are doing here.

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/DebuggerTransformer.java (hunk 2 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/DebuggerTransformer.java:631]: left over. this method is not called. removed

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/DebuggerTransformerTest.java (hunk 3 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/DebuggerTransformerTest.java:388]: done

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/SpanDecorationProbeInstrumentationTest.java (hunk 4 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/SpanDecorationProbeInstrumentationTest.java:341]: added test

**Pull request merged**